src/v/cluster/cloud_data/aggregated_log_reader.cc

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cluster/cloud_data/aggregated_log_reader.h"
+
+#include "bytes/iobuf.h"
+#include "bytes/iostream.h"
+#include "cloud_data/dl_placeholder.h"
+#include "cloud_data/types.h"
+#include "cloud_storage/cache_service.h"
+#include "cloud_storage/types.h"
+#include "cloud_storage_clients/types.h"
+#include "cluster/cloud_data/types.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "model/record_batch_reader.h"
+#include "model/record_batch_types.h"
+#include "model/record_utils.h"
+#include "storage/parser.h"
+#include "storage/record_batch_utils.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/io_priority_class.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/shared_ptr.hh>
+
+#include <boost/math/policies/error_handling.hpp>
+
+using namespace std::chrono_literals;
+
+namespace cluster::cloud_data {
+
+//
+struct hydrated_L0_object {
+    object_id id;
+    iobuf payload;
+};
+
+// Materialized placeholder extent
+struct placeholder_extent {
+    dl_placeholder placeholder;
+    ss::lw_shared_ptr<hydrated_L0_object> L0_object;
+};
+
+// Get dl_placeholder and the payload of the object and generate a record batch
+inline model::record_batch
+materialize_single_raft_data_batch(placeholder_extent& extent) {
+    auto offset = extent.placeholder.offset;
+    auto size = extent.placeholder.size_bytes;
+    vassert(
+      size() > model::packed_record_batch_header_size,
+      "L0 object is smaller ({}) than the batch header",
+      size());
+    auto header_bytes = extent.L0_object->payload.share(
+      offset(), model::packed_record_batch_header_size);
+    auto records_bytes = extent.L0_object->payload.share(
+      offset() + model::packed_record_batch_header_size,
+      size() - model::packed_record_batch_header_size);
+    // NOTE: the serialized raft_data batch doesn't have the offset set
+    // so we need to populate it from the placeholder batch. We also need
+    // to make sure that crc is correct.
+    auto header = storage::batch_header_from_disk_iobuf(
+      std::move(header_bytes));
+    model::record_batch batch(
+      header,
+      std::move(records_bytes),
+      model::record_batch::tag_ctor_ng{}); // TODO: fix compression
+    return batch;
+}
+
+/// Fetch data referenced by the placeholder batch and the content of the
+/// dl_placeholder
+inline ss::future<placeholder_extent> materialize_placeholder_extent(
```

Comment:
these don't seem to be invoked, so i'm guessing the consumer isn't fully working?